core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,780 @@
+    return findOvershadowedBy(rangeKeyOfGivenAug, aug.getMinorVersion(), fromState);
+  }
+
+  private List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> findOvershadowedBy(
+      RootPartitionRange rangeOfAug,
+      short minorVersion,
+      State fromState
+  )
+  {
+    Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> current = getStateMap(fromState)
+        .floorEntry(rangeOfAug);
+
+    if (current == null) {
+      return Collections.emptyList();
+    }
+
+    // Find the first key for searching for overshadowed atomicUpdateGroup
+    while (true) {
+      final Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> lowerEntry = getStateMap(fromState)
+          .lowerEntry(current.getKey());
+      if (lowerEntry != null && lowerEntry.getKey().startPartitionId == rangeOfAug.startPartitionId) {
+        current = lowerEntry;
+      } else {
+        break;
+      }
+    }
+
+    final List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> found = new ArrayList<>();
+    while (current != null && rangeOfAug.contains(current.getKey())) {
+      final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = current.getValue();
+      found.addAll(versionToGroup.subMap(versionToGroup.firstShortKey(), minorVersion).short2ObjectEntrySet());
+      current = getStateMap(fromState).higherEntry(current.getKey());
+    }
+    return found;
+  }
+
+  /**
+   * Handles addition of the atomicUpdateGroup to the given state
+   */
+  private void handleAdd(AtomicUpdateGroup<T> aug, State newStateOfAug)
+  {
+    if (newStateOfAug == State.STANDBY) {
+      // A standby atomicUpdateGroup becomes visible when its all segments are available.
+      if (aug.isFull()) {
+        // A visible atomicUpdateGroup becomes overshadowed when a fully available standby atomicUpdateGroup becomes
+        // visible which overshadows the current visible one.
+        findOvershadowedBy(aug, State.VISIBLE)
+            .forEach(entry -> transitPartitionChunkState(entry.getValue(), State.VISIBLE, State.OVERSHADOWED));
+        findOvershadowedBy(aug, State.STANDBY)
+            .forEach(entry -> transitPartitionChunkState(entry.getValue(), State.STANDBY, State.OVERSHADOWED));
+        transitPartitionChunkState(aug, State.STANDBY, State.VISIBLE);
+      }
+    }
+  }
+
+  private void addTo(AtomicUpdateGroup<T> aug, State state)
+  {
+    final AtomicUpdateGroup<T> existing = getStateMap(state)
+        .computeIfAbsent(RootPartitionRange.of(aug), k -> createMinorVersionToAugMap(state))
+        .put(aug.getMinorVersion(), aug);
+
+    if (existing != null) {
+      throw new ISE("AtomicUpdateGroup[%s] is already in state[%s]", aug, state);
+    }
+
+    handleAdd(aug, state);
+  }
+
+  public void add(PartitionChunk<T> chunk)
+  {
+    final PartitionChunk<T> existingChunk = knownPartitionChunks.put(chunk.getChunkNumber(), chunk);
+    if (existingChunk != null && !existingChunk.equals(chunk)) {
```

Comment:
Can you add a comment about when add might be called with a chunk that's already known? (e.g., handling replicas)

Reply:
This is just a sanity check to not allow different PartitionChunks to be added with the same id. This shouldn't happen even with replicas in the current code. Added a comment about this.